### PR TITLE
Feature/CO-1588-Implement-Integration-Tests

### PR DESCRIPTION
**Pull request by kalcicd**

### CODE REVIEW

**Review by Ruefa — CHANGES_REQUESTED** [review]
Nice work! I only left one comment. In the future you should practice committing more often. Commit early and often.

File: .gitignore [hunk 1]
```diff
@@ -92,4 +92,5 @@ env.bak/
 venv.bak/
 
 #VScode
-**/.vscode
+.vscode/
+.vscode/launch.json
```
Comment (tsoliangwu0130) [inline .gitignore]: Won't `.vscode/` ignore all of the files under `.vscode`?

**Review by Ruefa — APPROVED** [review]

File: tests/integration/integration-test.py [hunk 2]
```diff
@@ -7,6 +7,10 @@
 
 import utils
 
+err_obj = 'ErrorObject'
+park_res = 'ParkResource'
+owner_res = 'OwnerResource'
```
Comment (tsoliangwu0130) [inline tests/integration/integration-test.py]: If you want to define them as [constant variables](https://www.python.org/dev/peps/pep-0008/#constants), based on flake8/pep8 style guild, they must be all capital letters with underscores separating words.

> Constants are usually defined on a module level and written in all capital letters with underscores separating words. Examples include MAX_OVERFLOW and TOTAL.

File: tests/integration/integration-test.py [hunk 3]
```diff
@@ -34,73 +38,141 @@ def setup(cls, config_path, openapi_path):
...
-        nullable_fields = ['owner']
-        utils.test_endpoint(self, endpoint, 'PetResource', 200,
-                            nullable_fields=nullable_fields)
-
-    # Test case: GET /pets with species filter
-    def test_get_pets_with_filter(self, endpoint='/pets'):
-        testing_species = ['dog', 'CAT', 'tUrTlE']
+    # Test GET /parks
+    def test_get_all_parks(self, endpoint='/parks'):
+        utils.test_endpoint(self, endpoint, 'ParkResource', 200)
```
Comment (tsoliangwu0130) [inline tests/integration/integration-test.py]: Should `ParkResource` be `park_res`?

File: tests/integration/integration-test.py [hunk 4]
```diff
@@ -34,73 +34,145 @@ def setup(cls, config_path, openapi_path):
...
+            response = utils.test_endpoint(self, endpoint, 'ParkResource', 200,
+                                           query_params=params)
+            response_data = response.json()['data']
+            for resource in response_data:
+                actual_amenities = resource['attributes']['amenities']
+                for amenity in amenities.split(','):
+                    self.assertIn(amenity, actual_amenities)
+        for amenities in invalid_amenities:
+            params = {'filter[amenities][all]': amenities}
+            utils.test_endpoint(self, endpoint, 'ErrorObject', 400,
```
[the remark this answers is not in the document]
Comment (tsoliangwu0130) [reply tests/integration/integration-test.py]: Should this be `err_obj`?

File: tests/integration/integration-test.py [hunk 5]
```diff
@@ -34,73 +38,141 @@ def setup(cls, config_path, openapi_path):
...
+                actual_amenities = resource['attributes']['amenities']
+                self.assertTrue(
+                    # asserts if at least one item in amen_list is also in
+                    # actual_amenities
+                    any(True for x in actual_amenities if x in amen_list)
+                )
+        for amenities in invalid_amenities:
+            params = {'filter[amenities][all]': amenities}
+            utils.test_endpoint(self, endpoint, err_obj, 400,
+                                query_params=params)
```
Comment (tsoliangwu0130) [inline tests/integration/integration-test.py]: Should this be `filter[amenities][some]`? Besides the filter operation( `all` and `some`), the logic here is exactly the same as the `filter[amenities][all]`. Do you think we can use a helper function to combine them?

File: tests/integration/integration-test.py [hunk 6]
```diff
@@ -34,73 +38,141 @@ def setup(cls, config_path, openapi_path):
...
+            response = utils.test_endpoint(self, endpoint, park_res, 200,
+                                           query_params=params)
+            response_data = response.json()['data']
+            for resource in response_data:
+                actual_amenities = resource['attributes']['amenities']
+                self.assertTrue(
+                    # asserts if at least one item in amen_list is also in
+                    # actual_amenities
+                    any(True for x in actual_amenities if x in amen_list)
+                )
```
Comment (tsoliangwu0130) [inline tests/integration/integration-test.py]: Before the assertion, the logic is repeated with the test filter[amenities][all]. Do you think we can combine them in some ways?

File: tests/integration/integration-test.py [hunk 7]
```diff
@@ -34,73 +38,141 @@ def setup(cls, config_path, openapi_path):
...
-
-        for pet_id in invalid_pet_ids:
-            resource = 'ErrorObject'
-            utils.test_endpoint(self, f'{endpoint}/{pet_id}', resource, 404)
+                actual_state = resource['attributes']['location']['state']
+                self.assertEqual(actual_state, state)
+        for state in invalid_states:
+            params = {'filter[state]': state}
+            utils.test_endpoint(self, endpoint, 'ErrorObject', 400,
+                                query_params=params)
```
Comment (tsoliangwu0130) [inline tests/integration/integration-test.py]: The logic for each test cases are almost the same, again, I'd recommend thinking about how to combine them together and make your code [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

File: tests/integration/integration-test.py [hunk 8]
```diff
@@ -34,73 +38,141 @@ def setup(cls, config_path, openapi_path):
...
+        for owner_id in valid_owner_ids:
+            response = utils.test_endpoint(self, f'/owners/{owner_id}/parks',
+                                           'ParkResource', 200)
+            response_data = response.json()['data']
+            for park in response_data:
+                actual_id = park['relationships']['owner']['data']['id']
+                self.assertEqual(actual_id, owner_id)
+        for owner_id in invalid_owner_ids:
+            utils.test_endpoint(self, f'/owners/{owner_id}/parks', err_obj,
+                                404)
```
Comment (tsoliangwu0130) [inline tests/integration/integration-test.py]: This should be tested in `test_get_owner_by_id` (and you actually did, so this part is not actually needed.).

**Comment by tsoliangwu0130** [comment]
Good work, also don't forget to update your README.

**Review by tsoliangwu0130 — CHANGES_REQUESTED** [review]

File: tests/integration/integration-test.py [hunk 9]
```diff
@@ -34,73 +38,141 @@ def setup(cls, config_path, openapi_path):
...
+    # Test GET parks/{id}
+    def test_get_park_by_id(self):
+        valid_park_ids = self.test_cases['valid_park_ids']
+        invalid_park_ids = self.test_cases['invalid_park_ids']
+        for park_id in valid_park_ids:
+            response = utils.test_endpoint(self, f'/parks/{park_id}', park_res,
+                                           200)
+            actual_id = response.json()['data']['id']
+            self.assertEqual(actual_id, park_id)
+        for parkId in invalid_park_ids:
```
Comment (tsoliangwu0130) [inline tests/integration/integration-test.py]: Oh, I missed this one. We should use lowercase_snake_case for the [naming variable and function](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names) in Python. 

> Function names should be lowercase, with words separated by underscores as necessary to improve readability

**Review by tsoliangwu0130 — CHANGES_REQUESTED** [review]

File: tests/integration/utils.py [hunk 10]
```diff
@@ -289,6 +289,56 @@ def test_endpoint(self, endpoint, resource, response_code, query_params=None,
     return response
 
 
+def test_filter_params(self, endpoint, filter_type, valid_tests,
+                       invalid_tests):
+    PARK_RES = 'ParkResource'
```
Comment (tsoliangwu0130) [inline tests/integration/utils.py]: Not a big deal, but we usually put constant variables right after module import sections. You can just define this as a normal variable here.

File: tests/integration/utils.py [hunk 11]
```diff
@@ -289,6 +289,56 @@ def test_endpoint(self, endpoint, resource, response_code, query_params=None,
...
+            params = {f'filter[{filter_type}]': test}
+            response = test_endpoint(self, endpoint, PARK_RES, 200,
+                                     query_params=params)
+            response_data = response.json()['data']
+            self.assertFalse(response_data)
+    else:
+        for state in invalid_tests:
+            params = {f'filter[{filter_type}]': state}
+            test_endpoint(self, endpoint, 'ErrorObject', 400,
+                          query_params=params)
```
Comment (tsoliangwu0130) [inline tests/integration/utils.py]: Just a nitpick, but I think switch if and else will be easier to understand:

```python
if filter_type == 'state':
    for state in invalid_tests:
        do_something()
else:
    for test in invalid_tests:
        do_some_other_thing()
```

What do you think?

File: tests/integration/utils.py [hunk 12]
```diff
@@ -289,6 +289,56 @@ def test_endpoint(self, endpoint, resource, response_code, query_params=None,
...
+        params = {f'filter[amenities][{filter_type}]': amenities}
+        response = test_endpoint(self, endpoint, 'ParkResource', 200,
+                                 query_params=params)
+        response_data = response.json()['data']
+        for resource in response_data:
+            actual_amenities = resource['attributes']['amenities']
+            if filter_type == 'all':
+                for amenity in amen_list:
+                    self.assertIn(amenity, actual_amenities)
+            if filter_type == 'some':
```
Comment (tsoliangwu0130) [inline tests/integration/utils.py]: Could be `elif` to avoid unnecessary additional check.

**Review by tsoliangwu0130 — CHANGES_REQUESTED** [review]
I guess you missed my comment [here](https://github.com/osu-mist/parks-api/pull/17#issuecomment-536093448) 😛

File: tests/integration/utils.py (hunk 11 above)
Comment (kalcicd) [reply tests/integration/utils.py]: I agree. It makes it more clear that the test is just for filter[state]

**Review by tsoliangwu0130 — APPROVED** [review]

**Pull request merged**